modules/managers/galasa-managers-parent/galasa-managers-imstm-parent/dev.galasa.imstm.manager/src/main/java/dev/galasa/imstm/internal/ImstmDefaultLogonProvider.java

```diff
@@ -0,0 +1,133 @@
+            }
+
+            IImsSystem system = imsTerminal.getImsSystem();
+            imsTerminal.type(system.getZosImage().getVtamLogonString(system.getApplid())).enter().wfk();
+
+            waitForSignonScreen(imsTerminal);
+            logger.debug("Logged onto " + system);
+
+            if (imsTerminal.getLoginCredentialsTag().isEmpty()) {
+                throw new ImstmManagerException("No login credentials provided");
+            } else {
+                ICredentialsUsername creds = (ICredentialsUsername) this.cs.getCredentials(imsTerminal.getLoginCredentialsTag());
+
+                imsTerminal.positionCursorToFieldContaining("USERID:");
+                imsTerminal.tab();
+                imsTerminal.type(creds.getUsername());
+                if (creds instanceof ICredentialsUsernamePassword) {
+                    String pw = ((ICredentialsUsernamePassword) creds).getPassword();
+                    cts.registerText(pw, "Password for credential tag: " + imsTerminal.getLoginCredentialsTag());
+                    imsTerminal.positionCursorToFieldContaining("PASSWORD:");
+                    imsTerminal.tab();
+                    imsTerminal.type(pw);
+                    }
+                imsTerminal.enter().wfk();
+
+                waitForSignedOnText(imsTerminal);
+                logger.debug("Logged into IMS TM as user: " + creds.getUsername());
+            }
+
+            imsTerminal.clear().wfk();
+
+        } catch (Zos3270Exception | CredentialsException e) {
+            throw new ImstmManagerException("Problem logging onto the IMS system", e);
+        }
+
+        return true;
+    }
+
+    private void checkForInitialText(IImsTerminal imsTerminal) throws ImstmManagerException {
+        try {
+            imsTerminal.waitForTextInField(initialText);
+        } catch (Exception e) {
+            throw new ImstmManagerException(
+                    "Unable to logon to IMS, initial screen does not contain '" + initialText + "'");
+        }
+    }
+
+    private void waitForSignonScreen(IImsTerminal imsTerminal) throws ImstmManagerException {
+        try {
+            imsTerminal.waitForTextInField("DFS3649A");
+        } catch (Exception e) {
+            throw new ImstmManagerException("Unable to wait for the initial IMS screen, looking for 'DFS3649I'",
+                    e);
+        }
+    }
+
+    private void waitForSignedOnText(IImsTerminal imsTerminal) throws ImstmManagerException {
+
+        String[] pass = { "DFS3650I" };
+        String[] fail = { "REJECTED" };  // Any signon failure will give us DFS3649A again, with REJECTED and a reason code.
+                                         // However, we can't search for DFS3649A as it's already on the screen.
+
+        try {
+            imsTerminal.waitForTextInField(pass, fail);
+        } catch (Exception e) {
+            throw new ImstmManagerException("Unable to sign on, looking for '" + String.join("', '", pass) + "'",
```

Comment:
Don't understand why you are using String.join here... pass is a 1-element array, you could use pass[0] instead ?

Use `MessageFormat.format("Unable to sign on, looking for ''{0}''",pass[0]);` instead ?

Reply:
Does the error get logged ? Might be good to log the exception prior to throwing it, or get the exception constructor to log ?